[Profiler] Add `disabled` and `global` methods to ProfilerConfig.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83965
* #83964
* #83895
* #83894
* #83893
* #83892
* __->__ #83891
* #83890
* #83963
* #83622

`ProfilerState::Disabled` and `ProfilerState::KINETO_ONDEMAND` have special semantics. The former is somewhat intuitive, but the degree of behavior branching on the latter (and why the branching is necessary) is less clear. By factoring the enum checks into methods, we can both clairify intent and future proof in case we ever add other global profiling contexts.

Differential Revision: [D38917980](https://our.internmc.facebook.com/intern/diff/D38917980/)

@diff-train-skip-merge

Commit:
[Profiler] Add `disabled` and `global` methods to ProfilerConfig.

`ProfilerState::Disabled` and `ProfilerState::KINETO_ONDEMAND` have special semantics. The former is somewhat intuitive, but the degree of behavior branching on the latter (and why the branching is necessary) is less clear. By factoring the enum checks into methods, we can both clairify intent and future proof in case we ever add other global profiling contexts.

Differential Revision: [D38917980](https://our.internmc.facebook.com/intern/diff/D38917980/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `disabled` and `global` methods to ProfilerConfig."


`ProfilerState::Disabled` and `ProfilerState::KINETO_ONDEMAND` have special semantics. The former is somewhat intuitive, but the degree of behavior branching on the latter (and why the branching is necessary) is less clear. By factoring the enum checks into methods, we can both clairify intent and future proof in case we ever add other global profiling contexts.

Differential Revision: [D38917980](https://our.internmc.facebook.com/intern/diff/D38917980/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `disabled` and `global` methods to ProfilerConfig."


`ProfilerState::Disabled` and `ProfilerState::KINETO_ONDEMAND` have special semantics. The former is somewhat intuitive, but the degree of behavior branching on the latter (and why the branching is necessary) is less clear. By factoring the enum checks into methods, we can both clairify intent and future proof in case we ever add other global profiling contexts.

Differential Revision: [D38917980](https://our.internmc.facebook.com/intern/diff/D38917980/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Add `disabled` and `global` methods to ProfilerConfig."


`ProfilerState::Disabled` and `ProfilerState::KINETO_ONDEMAND` have special semantics. The former is somewhat intuitive, but the degree of behavior branching on the latter (and why the branching is necessary) is less clear. By factoring the enum checks into methods, we can both clairify intent and future proof in case we ever add other global profiling contexts.

Differential Revision: [D38917980](https://our.internmc.facebook.com/intern/diff/D38917980/)

[ghstack-poisoned]

slgong-fb commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -558,13 +558,18 @@ void enableProfiler(
 
-  if (config.state == ProfilerState::KINETO ||
-      config.state == ProfilerState::KINETO_GPU_FALLBACK) {
+  if (config.global()) {
```
Should this line be like below? `ProfilerState::KINETO` seems to be missed.
```
if (config.state == ProfilerState::KINETO || config.global()) {
```

robieta replied:
No. `config.global()` is on-demand. `ProfilerState::KINETO` and `ProfilerState::KINETO_GPU_FALLBACK` are covered by the else branch.

slgong-fb reviewed (approved):
LGTM except for one 'if' statement (comment added).

Commit:
Update on "[Profiler] Add `disabled` and `global` methods to ProfilerConfig."


`ProfilerState::Disabled` and `ProfilerState::KINETO_ONDEMAND` have special semantics. The former is somewhat intuitive, but the degree of behavior branching on the latter (and why the branching is necessary) is less clear. By factoring the enum checks into methods, we can both clairify intent and future proof in case we ever add other global profiling contexts.

Differential Revision: [D38917980](https://our.internmc.facebook.com/intern/diff/D38917980/)

[ghstack-poisoned]

robieta commented:
@pytorchbot merge

weiwangmeta commented:
@pytorchmergebot merge -L

weiwangmeta commented:
@pytorchbot merge -l
